projects/dev/src/app/group/group-manage/pending-request/pending-request.component.ts

```diff
@@ -22,87 +45,141 @@ export class PendingRequestComponent implements OnInit, OnChanges {
     { field: "at", order: -1 },
     { field: "member_id", order: 1 },
   ];
-  prevSortMeta = "-at member_id";
-
-  @Output() onAccept = new EventEmitter<any>();
-  @Output() onReject = new EventEmitter<any>();
-
-  groupSwitch = [
-    {
-      label: "This group only",
-    },
-    {
-      label: "All subgroups",
-    },
-  ];
+  prevSortMeta: string[] = GROUP_REQUESTS_API.sort;
+
+  requestAction: AcceptReject = AcceptReject.None;
+  selection = [];
 
-  _setRequestData(reqs: PendingRequest[]) {
-    this.requests = [];
-
-    for (const req of reqs) {
-      this.requests.push({
-        member_id: req.member_id,
-        "joining_user.login": `${req.joining_user.first_name || ""} ${
-          req.joining_user.last_name || ""
-        } (${req.joining_user.login || ""})`,
-        "grade": req.joining_user && req.joining_user.grade ? req.joining_user.grade : null,
-        at: req.at,
+  _reloadData() {
+    this.groupService
+      .getManagedRequests(this.id, this.prevSortMeta)
+      .subscribe((reqs: PendingRequest[]) => {
+        this.requests = reqs.map((req) => {
+          return {
+            member_id: req.member_id,
+            joining_user: req.joining_user,
+            at: req.at,
+          };
+        });
       });
+  }
+
+  _handleActionResponse(result: RequestActionResponse, verb: string, msg: string) {
+    if (result.success === true && result.message === "updated") {
```

Comment:
Else... fail silently? 🤔

Reply:
The else case is the error processing as the api returns error code so it automatically goes to `_processRequestError` method.
What do you suggest?

Reply:
I am sure I can find other case that will fail silently (= bug difficult to find)...
Just if the message is not "updated" (whatever the reason is), it will fail silently. Actually, if the success is false as well but the status code 200. Ok we could consider them as bugs from the backend, but you cannot just ignore them! 
So either, you consider them as errors and you transform them to an errors for the observers... or you ignore the errors and you threat them as success.  Probably option 1 is better.
When you develop on UI, you have to adopt defensive programming, your UI has to react properly whatever the input is. For instance, that's why we used `return ["success", "unchanged"].includes(status);` and not `return status !== "error"` or doing a switch-case on all values. A value may be something unexpected... and here we'll just consider it as a failure, so it is handled.

Reply:
That makes sense.
fixed.

Reply:
Mmh, you do this same check twice now, it is worse. 😕

Of course, `_displayResponseToast` should not deal about the whole response but just the id->result dictionary only.

Grr, it makes me realize you do not even validate the type of the "data" attribute of the response... 😔   <copy paste the paragraph about defensive programming>

Reply:
fixed.

Reply:
Fixed? My last remark is still there at least.

Reply:
What's your suggestion?

Reply:
For the data, it is not any object, you expect a dictionary of ids (string) to result (string). Apparently you have the "Record" type in TS, so you can say you expect the object to be a  `Record<string, string>` (directly in the object mapping). It will 

👍  for mergeMap, probably you can do it in the service directly (and later, you'll be able to re-use something similar in the other services).
BTW, a simple `tap` here would more simple
```
.pipe(
      tap(res => {
        if (res.success === false || res.message !== "updated" || typeof res.data !== "object") {
          throw "Unknown error";
        }
      })
    )```


How have you tested that these kind of behavior work as you expect?

Reply:
I changed the `res.success === false` to `res.success === true` and tested and it threw the error and displayed the toast as I expected.

Reply:
You haven't address the part of the comment related to 
> For the data, it is not any object, you expect a dictionary of ids (string) to result (string). Apparently you have the "Record" type in TS, so you can say you expect the object to be a Record<string, string> (directly in the object mapping). It will

Reply:
Ah, yes, I will fix it.

Reply:
By the way, what should I validate with the `res.data`?
I only added `!res.data` but what else can I add to its validation?